lib/scripts.js

```diff
@@ -143,20 +143,23 @@ var scripts = {
         return job.queue.toKey(name);
       }
     );
-
     keys.push(job.lockKey());
 
-    //
-    // INVESTIGATE: Should'nt we check if we have the lock before trying to move the
-    // job?
-    //
+    var deleteJob = 'redis.call("DEL", KEYS[3])';
+    var moveJob = [
+      'redis.call("SADD", KEYS[2], ARGV[1])',
+      'redis.call("HSET", KEYS[3], "returnvalue", ARGV[2])',
+    ].join('\n');
+
     var script = [
-      'if redis.call("EXISTS", KEYS[3]) == 1 then',
-      ' redis.call("SADD", KEYS[2], ARGV[1])',
-      ' redis.call("HSET", KEYS[3], "returnvalue", ARGV[2])',
-      ' redis.call("LREM", KEYS[1], 0, ARGV[1])',
-      ' if redis.call("get", KEYS[4]) == ARGV[3] then', // Remove the lock
-      '  return redis.call("del", KEYS[4])',
+      'if redis.call("EXISTS", KEYS[3]) == 1 then', // Make sure job exists
+      ' if redis.call("GET", KEYS[4]) == ARGV[3] then', // Makes sure we own the lock
+      '  redis.call("LREM", KEYS[1], 0, ARGV[1])',
+      removeOnComplete ? deleteJob : moveJob,
+      '  redis.call("DEL", KEYS[4])',
+      '  return 0',
+      ' else',
+      '  return -1',
       ' end',
```

Comment:
A slight issue here, you need to name differently the script when you specify removeOnComplete, otherwise, since scripts are cached, if you do not use removeOnComplete in a future instance, it will use the old cached one. I know this is probably not going to happen, but for the sake of formality :)